### PR TITLE
Ravi kiran@team yellow

### PR DESCRIPTION
**Pull request by trkiran777**
Please check the files.

### CODE REVIEW

File: contact_manager.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+class Contacts:
+
+    provider = {'Airtel': ['9900', '9800', '9811'], 'BSNL': ['9440', '9822'], 'Idea': ['9848', '9912'],
+                'Reliance': ['9300', '9812']}
+
+    def __init__(self, data):
+        self.data = data
+
+    def add_person(self, contact):
```
Comment (boyravikumar) [inline contact_manager.py]: add_contact

File: contact_manager.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
+import json
+
+
+class Address:
+
+    def __init__(self, street, city, state, pin_code):
```
Comment (boyravikumar) [inline contact_manager.py:4]: getters and setters for all classes!!

File: contact_manager.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        self.address = Address(street, city, state, pin_code)
+
+
+class Contacts:
+
+    provider = {'Airtel': ['9900', '9800', '9811'], 'BSNL': ['9440', '9822'], 'Idea': ['9848', '9912'],
+                'Reliance': ['9300', '9812']}
+
+    def __init__(self, data):
+        self.data = data
```
Comment (boyravikumar) [inline contact_manager.py]: contact_list

File: contact_manager.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                self.data[phone_no].addres.city = fields[key]
+            elif key == 'state':
+                self.data[phone_no].addres.state = fields[key]
+            elif key == 'pin_code':
+                    self.data[phone_no].addres.pin_code = fields[key]
+
+    def delete_person(self, phone_no):
+        del self.data[phone_no]
+
+    def get_details(self, phone_no):
```
Comment (boyravikumar) [inline contact_manager.py]: get_contact

File: contact_manager.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    def __init__(self, data):
+        self.data = data
+
+    def add_person(self, contact):
+        self.data[contact.phone_no] = contact
+
+    def modify_person(self, phone_no, fields):
+        for key in fields.keys():
+            if key == 'name':
```
Comment (boyravikumar) [inline contact_manager.py]: make all constants class called ContactDetails and use it!!

File: contact_manager.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+                         value["address"]["city"], value["address"]["state"], value["address"]["pin_code"])
+            data[key] = ct
+        c = Contacts(data)
+        print('1: ADD_PERSON, 2: MODIFY_PERSON, 3: DELETE_PERSON, 4: DISPLAY_DETAILS, 5: DISPLAY_PROVIDER,'
+              ' 6: DISPLAY_RECORDS_OF_PROVIDER, 7: DISPLAY_RECORDS_CONTAINING_STRING, 8:EXIT')
+        while True:
+            print('Enter your command:')
+            ip = input()
+            if ip == '1':
+                name = input('Name:')
```
Comment (boyravikumar) [inline contact_manager.py]: divide in to functions!!

File: contact_manager.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    provider = {'Airtel': ['9900', '9800', '9811'], 'BSNL': ['9440', '9822'], 'Idea': ['9848', '9912'],
+                'Reliance': ['9300', '9812']}
+
+    def __init__(self, data):
+        self.data = data
+
+    def add_person(self, contact):
+        self.data[contact.phone_no] = contact
+
+    def modify_person(self, phone_no, fields):
```
Comment (boyravikumar) [inline contact_manager.py]: make use of get contact and update that contact!!

File: contact_manager.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+            elif key == 'pin_code':
+                    self.data[phone_no].addres.pin_code = fields[key]
+
+    def delete_person(self, phone_no):
+        del self.data[phone_no]
+
+    def get_details(self, phone_no):
+        return self.data[phone_no]
+
+    def get_provider(self, phone_no):
```
Comment (boyravikumar) [inline contact_manager.py:130]: Create a provider Class!! Provide Manager!! Initialize all provider and ProviderManager!! Checking for aprovider should happen in ProviderManager!!

File: contact_manager.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        s = phone_no[0:4]
+        for key, values in Contacts.provider.items():
+            if s in values:
+                return key
+        return 'Others'
+
+    def get_records_of_provider(self, provider_name):
```
Comment (boyravikumar) [inline contact_manager.py]: get_contacts_by_provider

File: contact_manager.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        li = []
+        for key, values in self.data.items():
+            if key[0:4] in Contacts.provider[provider_name]:
+                li.append(values)
+        return li
+
+    def get_records_containing_string(self, st, field):
```
Comment (boyravikumar) [inline contact_manager.py]: think of some good name!!

File: contact_manager.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+            ip = input()
+            if ip == '1':
+                name = input('Name:')
+                phone_no = input('Phone:')
+                email = input('Email:')
+                street = input('Street:')
+                city = input('City:')
+                state = input('State:')
+                pin_code = input('Pin_code:')
+                if phone_no not in c.data.keys() and len(phone_no) == 10 and phone_no[0] != '0' \
```
Comment (boyravikumar) [inline contact_manager.py]: expose a method to check if contact pr contact_number already exists

Comment (boyravikumar) [inline contact_manager.py]: Move the validations to respective class and make use of them here!!

File: contact_manager.py [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+
+            elif ip == '7':
+                fields = ['name', 'phone_no', 'email', 'street', 'city', 'state', 'pin_code']
+                st = input('Enter a string:')
+                field = input('Enter a field:')
+                if field in fields:
+                    li = c.get_records_containing_string(st, field)
+                    for item in li:
+                        print('Name: {name}, Phone_no: {phone_no}, Email: {email}, Street: {street}, City: {city}, '
+                              'State: {state}, Pin_code: {pin_code}'.format(name=item.name,
```
Comment (boyravikumar) [inline contact_manager.py]: expose a method call get_details_string / get_formatted_details and call it here!!

File: contact_manager.py [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+                                                                            phone_no=item.phone_no,
+                                                                            email=item.email,
+                                                                            street=item.address.street,
+                                                                            city=item.address.city,
+                                                                            state=item.address.state,
+                                                                            pin_code=item.address.pin_code))
+            elif ip == '8':
+                final_data = {}
+                for key, value in c.data.items():
+                    final_data[key] = value.__dict__
```
Comment (boyravikumar) [inline contact_manager.py]: don't use **dict**!! 
expose a method cal get_dict or get_json in Contact (which will make use of get_json of Address) and make use of it.

File: contact_manager.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+                        continue
+                else:
+                    print('Phone number is not valid...Please check it.')
+            elif ip == '2':
+                phone_no = input('Enter Phone_no:')
+                if phone_no in c.data.keys():
+                    n = int(input('Enter no. of fields to change:'))
+                    d = {}
+                    for i in range(n):
+                        key = input('Field:')
```
Comment (boyravikumar) [inline contact_manager.py]: validate the fields taken!!

File: contact_manager.py [hunk 15]
```diff
@@ -1,218 +1,281 @@
...
+
+
+class Contacts:
+    def __init__(self, contact_list):
+        self.contact_list = contact_list
+
+    def add_contact(self, contact):
+        self.contact_list[contact.phone_no] = contact
+
+    def modify_contact(self, phone_no, fields):
```
Comment (boyravikumar) [inline contact_manager.py:106]: get_contact and make modifications!!

File: contact_manager.py [hunk 16]
```diff
@@ -1,218 +1,281 @@
...
+    PHONE_NO = 'phone_no'
+    EMAIL = 'email'
+    ADDRESS = 'address'
+    STREET = 'street'
+    CITY = 'city'
+    STATE = 'state'
+    PIN_CODE = 'pin_code'
+
+ 
+class Provider:
```
Comment (boyravikumar) [inline contact_manager.py]: change the Provider class and make changes as discussed!!

File: contact_manager.py [hunk 17]
```diff
@@ -1,218 +1,281 @@
...
 
-    def __init__(self, data):
-        self.data = data
+class ProviderManager:
+    def __init__(self):
+        self.provider = Provider()
 
-    def add_person(self, contact):
-        self.data[contact.phone_no] = contact
+    def set_provider_name(self, phone_no):
```
Comment (boyravikumar) [inline contact_manager.py]: This method is not needed!!

File: contact_manager.py [hunk 18]
```diff
@@ -1,218 +1,281 @@
...
+    def get_contact(self, phone_no):
+        return self.contact_list[phone_no]
 
     def get_provider(self, phone_no):
-        s = phone_no[0:4]
-        for key, values in Contacts.provider.items():
-            if s in values:
-                return key
-        return 'Others'
+        ProviderManager.get_provider_name(phone_no[0:4])
```
Comment (boyravikumar) [inline contact_manager.py:131]: read clearly difference between static and non static methods!! and their usage!!

File: contact_manager.py [hunk 19]
```diff
@@ -1,218 +1,281 @@
...
+            if key[0:4] in p[provider_name]:
                 li.append(values)
         return li
 
-    def get_records_containing_string(self, st, field):
+    def get_contacts_by_string(self, st, field):
         li = []
-        for key, values in self.data.items():
-            if field == 'name' and st in values.name:
+        for key, values in self.contact_list.items():
```
Comment (boyravikumar) [inline contact_manager.py]: rename to phone and contact

File: contact_manager.py [hunk 20]
```diff
@@ -1,218 +1,281 @@
...
         return li
 
 
 class ContactManager:
+    contact_list = {}
+    contacts = Contacts(contact_list)
 
-    def manage_contacts(self):
-        data = {}
+    def get_contacts_from_json(self):
```
Comment (boyravikumar) [inline contact_manager.py]: load_contacts_from_file

File: contact_manager.py [hunk 21]
```diff
@@ -1,218 +1,281 @@
...
+        fields = [ContactDetails.NAME, ContactDetails.PHONE_NO, ContactDetails.EMAIL, ContactDetails.STREET,
+                  ContactDetails.CITY, ContactDetails.STATE, ContactDetails.PIN_CODE]
+        st = input('Enter a string:')
+        field = input('Enter a field:')
+        if field in fields:
+            li = cm.contacts.get_contacts_by_string(st, field)
+            for item in li:
+                print(cm.get_formatted_details(item))
+
+    def add_contact_list_to_json(self):
```
Comment (boyravikumar) [inline contact_manager.py]: dump/save contact to file!!

File: Contacts.py [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+        return self.contact_list[phone_no]
+
+    def get_provider(self, phone_no):
+        ProviderManager.get_provider_name(phone_no[0:4])
+
+    def get_contacts_by_provider(self, provider_name):
+        li = []
+        p = Provider()
+        p = p.__dict__
+        for key, values in self.contact_list.items():
```
Comment (boyravikumar) [inline Contacts.py]: change it according to the new Provider / Manage class

File: Contacts.py [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+        return 'Name: {name}, Phone_no: {phone_no}, Email: {email}, Street: {street}, City: {city}, ' \
+               'State: {state}, Pin_code: {pin_code}'.format(name=details.name,
+                                                             phone_no=details.phone_no,
+                                                             email=details.email,
+                                                             street=details.address.street,
+                                                             city=details.address.city,
+                                                             state=details.address.state,
+                                                             pin_code=details.address.pin_code)
+
+    def check_contact_exists_or_not(self, phone_no):
```
Comment (boyravikumar) [inline Contacts.py]: rename to is_contact_exists

File: Contacts.py [hunk 24]
```diff
@@ -0,0 +1,291 @@
...
+    def add_contact(self):
+        name = input('Name:')
+        phone_no = input('Phone:')
+        email = input('Email:')
+        street = input('Street:')
+        city = input('City:')
+        state = input('State:')
+        pin_code = input('Pin_code:')
+        contact_exist = cm.check_contact_exists_or_not(phone_no)
+        if (not contact_exist) and len(phone_no) == 10 and phone_no[0] != '0' \
```
Comment (boyravikumar) [inline Contacts.py]: Create a static method isValidPhoneNo is Contact and make use of it here!!

File: Contacts.py [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+            elif option == '2':
+                return
+        else:
+            print('Phone number is not valid...Please check it.')
+
+    def modify_contact(self):
+        fields = [ContactDetails.NAME, ContactDetails.PHONE_NO, ContactDetails.EMAIL, ContactDetails.STREET,
+                  ContactDetails.CITY, ContactDetails.STATE, ContactDetails.PIN_CODE]
+        phone_no = input('Enter Phone_no:')
+        if phone_no in cm.contacts.contact_list.keys():
```
Comment (boyravikumar) [inline Contacts.py]: use is_existing_contact!!

File: Contacts.py [hunk 26]
```diff
@@ -0,0 +1,291 @@
+import json
+
+
+class Address:
+    def __init__(self, street, city, state, pin_code):
+        self.street = street
```
Comment (boyravikumar) [inline Contacts.py]: getters_setters!!

File: Contacts.py [hunk 27]
```diff
@@ -0,0 +1,291 @@
...
+    def get_contacts_by_provider(self, provider_name):
+        li = []
+        p = Provider()
+        p = p.__dict__
+        for key, values in self.contact_list.items():
+            if key[0:4] in p[provider_name]:
+                li.append(values)
+        return li
+
+    def get_contacts_by_string(self, st, field):
```
Comment (boyravikumar) [inline Contacts.py]: get_contacts_by_field

File: Contacts.py [hunk 28]
```diff
@@ -0,0 +1,291 @@
...
+        for item in records:
+            print(cm.get_formatted_details(item))
+
+    def get_contacts_by_string(self):
+        fields = [ContactDetails.NAME, ContactDetails.PHONE_NO, ContactDetails.EMAIL, ContactDetails.STREET,
+                  ContactDetails.CITY, ContactDetails.STATE, ContactDetails.PIN_CODE]
+        st = input('Enter a string:')
+        field = input('Enter a field:')
+        if field in fields:
+            li = cm.contacts.get_contacts_by_string(st, field)
```
Comment (boyravikumar) [inline Contacts.py]: No Data Found!!

File: contact_manager.py (hunk 15 above)
Comment (trkiran777) [inline contact_manager.py:106]: Ravi kiran ...

File: contact_manager.py (hunk 16 above)
Comment (trkiran777) [inline contact_manager.py]: Sanjay...

File: contact_manager.py (hunk 17 above)
Comment (trkiran777) [inline contact_manager.py]: Sanjay

File: contact_manager.py (hunk 20 above)
Comment (trkiran777) [inline contact_manager.py]: Sri hari

File: contact_manager.py (hunk 21 above)
Comment (trkiran777) [inline contact_manager.py]: Sri hari

File: Contacts.py (hunk 22 above)
Comment (trkiran777) [inline Contacts.py]: Sanjay...

File: Contacts.py (hunk 24 above)
Comment (trkiran777) [inline Contacts.py]: Ravi kiran

File: Contacts.py (hunk 25 above)
Comment (trkiran777) [inline Contacts.py]: Ravi kiran

File: Contacts.py (hunk 26 above)
Comment (trkiran777) [inline Contacts.py]: Sanjay

**Pull request closed without merging**